feat: support system to bridge txs

marshacb opened the pull request:
### Description
This PR introduces the `SystemToBridge` tx direction for constructing BridgeHub transactions from AssetHub.

#### Changes
* adds support for the `transferAssetsUsingTypeAndThen` extrinsic.
* adds `SystemToBridge` direction. 
* In addition to allowing Parachain IDs, the API now allows for a `GlobalConsensus` location (e.g. `{"parents":"2","interior":{"X1":{"GlobalConsensus":{"Ethereum":{"chainId":"11155111"}}}}}`) to be provided as a value for the`destChainId` parameter in `createTransferTransaction`. If a destChainId is provided containing a known `X1` `GlobalConsensus` junction it will be considered a Bridge tx. Currently known consensus junctions are set to `polkadot`, `kusama`, `westend`, `rococo`, and `ethereum`.
* Allows for registered global consensus asset locations on AssetHub (e.g. `{"parents":"2","interior":{"X2":[{"GlobalConsensus":{"Ethereum":{"chainId":"11155111"}}},{"AccountKey20":{"network":null,"key":"0xfff9976782d46cc05630d1f6ebab18b2324d6b14"}}]}}` on AssetHub Rococo) to be provided as `assetIds`  in `createTransferTransaction`.
* Adds error validation checks for `SystemToBridge` direction
* Adds unit and integration tests for `SystemToBridge` direction
* Removed unnecessary check restricting `generalKey` locations to XCM version 2 in `resolveMultiLocation`

### BridgeHub Snowbridge: (Rococo AssetHub WETH -> Ethereum Sepolia)
Example of constructing a `transferAssets` call sending Snowbridge `WETH` to Ethereum Sepolia from Rococo AssetHub:
```ts
	const { api, specName, chainName, safeXcmVersion } = await constructApiPromise(
		'wss://rococo-asset-hub-rpc.polkadot.io',
	);
	const assetApi = new AssetTransferApi(api, specName, safeXcmVersion, {
		chainName,
	});

	let callInfo = await assetApi.createTransferTransaction(
			`{"parents":"2","interior":{"X1":{"GlobalConsensus":{"Ethereum":{"chainId":"11155111"}}}}}`,
			'0x6E733286C3Dc52C67b8DAdFDd634eD9c3Fb05B5B',
			[
				`{"parents":"2","interior":{"X2":[{"GlobalConsensus":{"Ethereum":{"chainId":"11155111"}}},{"AccountKey20":{"network":null,"key":"0xfff9976782d46cc05630d1f6ebab18b2324d6b14"}}]}}`,
			],
			['1000000000000'],
			{
				format: 'call',
				xcmVersion: 4,
			},
		);

		console.log(callInfo);
```

Expected Output
```
{
  origin: 'statemine',
  dest: 'ethereum',
  direction: 'SystemToBridge',
  xcmVersion: 4,
  method: 'transferAssets',
  format: 'call',
  tx: '0x1f0b04020109079edaa80204000103006e733286c3dc52c67b8dadfdd634ed9c3fb05b5b0404020209079edaa8020300fff9976782d46cc05630d1f6ebab18b2324d6b1400070010a5d4e80000000000'
}

The following decoded tx:
 {
    "args": {
        "dest": {
            "V4": {
                "parents": "2",
                "interior": {
                    "X1": [
                        {
                            "GlobalConsensus": {
                                "Ethereum": {
                                    "chainId": "11,155,111"
                                }
                            }
                        }
                    ]
                }
            }
        },
        "beneficiary": {
            "V4": {
                "parents": "0",
                "interior": {
                    "X1": [
                        {
                            "AccountKey20": {
                                "network": null,
                                "key": "0x6e733286c3dc52c67b8dadfdd634ed9c3fb05b5b"
                            }
                        }
                    ]
                }
            }
        },
        "assets": {
            "V4": [
                {
                    "id": {
                        "parents": "2",
                        "interior": {
                            "X2": [
                                {
                                    "GlobalConsensus": {
                                        "Ethereum": {
                                            "chainId": "11,155,111"
                                        }
                                    }
                                },
                                {
                                    "AccountKey20": {
                                        "network": null,
                                        "key": "0xfff9976782d46cc05630d1f6ebab18b2324d6b14"
                                    }
                                }
                            ]
                        }
                    },
                    "fun": {
                        "Fungible": "1,000,000,000,000"
                    }
                }
            ]
        },
        "fee_asset_item": "0",
        "weight_limit": "Unlimited"
    },
    "method": "transferAssets",
    "section": "polkadotXcm"
}
```

### BridgeHub Polkadot<>Kusama Bridge: (Kusama AssetHub KSM -> Polkadot AssetHub)
Example of constructing a `transferAssets` call sending KSM  to Polkadot AssetHub from Kusama AssetHub:
```ts
const { api, specName, chainName, safeXcmVersion } = await constructApiPromise(
		'wss://kusama-asset-hub-rpc.polkadot.io',
	);
	const assetApi = new AssetTransferApi(api, specName, safeXcmVersion, {
		chainName,
	});

let callInfo = await assetApi.createTransferTransaction(
			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Polkadot"}}}`,
			'5EWNeodpcQ6iYibJ3jmWVe85nsok1EDG8Kk3aFg8ZzpfY1qX',
			[`{"parents":"1","interior":{"Here":""}}`],
			['1000000000000'],
			{
				format: 'call',
				xcmVersion: 4,
			},
		);

		console.log(callInfo);
```

Expected Output:
```
{
  origin: 'statemine',
  dest: 'polkadot',
  direction: 'SystemToBridge',
  xcmVersion: 4,
  method: 'transferAssets',
  format: 'call',
  tx: '0x1f0b040202090200a10f04000101006c0c32faf970eacb2d4d8e538ac0dab3642492561a1be6f241c645876c056c1d04040100000700e87648170000000000'
}

The following decoded tx:
 {
    "args": {
        "dest": {
            "V4": {
                "parents": "2",
                "interior": {
                    "X2": [
                        {
                            "GlobalConsensus": "Polkadot"
                        },
                        {
                            "Parachain": "1,000"
                        }
                    ]
                }
            }
        },
        "beneficiary": {
            "V4": {
                "parents": "0",
                "interior": {
                    "X1": [
                        {
                            "AccountId32": {
                                "network": null,
                                "id": "0x6c0c32faf970eacb2d4d8e538ac0dab3642492561a1be6f241c645876c056c1d"
                            }
                        }
                    ]
                }
            }
        },
        "assets": {
            "V4": [
                {
                    "id": {
                        "parents": "1",
                        "interior": "Here"
                    },
                    "fun": {
                        "Fungible": "100,000,000,000"
                    }
                }
            ]
        },
        "fee_asset_item": "0",
        "weight_limit": "Unlimited"
    },
    "method": "transferAssets",
    "section": "polkadotXcm"
}
```

### Note
* If a given asset location doesn’t exist in the `ForeignAssets` pallet of the origin chain the API will throw an `AssetNotFound` error
* Currently assumes the user is sending an asset that is known/valid from the context of the destination chain (checks that the asset exists on chain for the origin)
* Providing the `assetTransferType` option will resolve the call to `transferAssetsUsingTypeAndThen` if it is found in the runtime. When providing a value of `RemoteReserve` for `assetTransferType` or `feesTransferType`, a location value is expected for `remoteReserveAssetTransferTypeLocation` and/or `remoteReserveFeesTransferTypeLocation` respectively.

related issue: [#382](https://github.com/paritytech/asset-transfer-api/issues/382)
closes: [#400](https://github.com/paritytech/asset-transfer-api/issues/400)

IkerAlus commented:
great to see this out already!

One request @marshacb : once this is ready for review, it would be really helpful if you add details to the original post of the differences in UX this feature implies (for example, destinations as multilocations and so on)

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/parity-tech-update-for-march/7226/1

acatangiu reviewed (commented):
Output/result looks good 👍 

Does it support P<>K bridge as well?

marshacb commented:
> Output/result looks good 👍
> 
> Does it support P<>K bridge as well?

Hey @acatangiu, thanks for checking this out! Yes, this feature will support P<>K bridge transaction construction as well. 
Just to note, this is currently being reworked to support the newer `transfer_assets_using_type_and_then` call for bridge transfer cases where `AssetHub` is required to be treated as the remote reserve.

IkerAlus reviewed (changes requested):
Couple of general comments:

- Can we add an example where we send (bridge) the native token of another ecosystem back to its native ecosystem? For example ROC from Westend AH back to Rococo AH?
- Any reason in particular to have the example with Trappist? In principle, it is normal foreign asset transfer from AH which is already covered in other examples.

bee344 reviewed (approved):
Two nits but great work

bee344 commented on README.md:
```diff
@@ -31,19 +31,20 @@ The below chart is focusing on what directions are supported for constructing as
 | System to Relay        |         ✅         |      ✅            |          ✅         |
+| System to System       |         ✅         |      ✅            |          ✅         |
+| System to Bridge       |         ❌         |      ✅            |          ✅         |
+| System to Parachain    |         ✅         |      ✅            |          ✅		  |
```
```suggestion
| System to Parachain    |         ✅         |      ✅            |          ✅		|
```

bee344 commented on README.md:
```diff
@@ -210,6 +211,32 @@ interface TransferArgsOpts<T extends Format> {
 	 * Default is false.
 	 */
 	transferLiquidToken?: boolean;
+		/**
```
```suggestion
	/**
```

marshacb commented:
> Couple of general comments:
> 
> * Can we add an example where we send (bridge) the native token of another ecosystem back to its native ecosystem? For example ROC from Westend AH back to Rococo AH?
> * Any reason in particular to have the example with Trappist? In principle, it is normal foreign asset transfer from AH which is already covered in other examples.

* I added examples of bridging DOT back to Polkadot from Kusama (and a similar one for Kusama).
* No reason in particular so I removed the Trappist example

IkerAlus commented on examples/rococo/assetHub/foreignAssetTransfers/teleport/assetHubWETHToBridgeHub.ts:
```diff
@@ -0,0 +1,53 @@
+/**
+ * When importing from @substrate/asset-transfer-api it would look like the following
```
Better to remove this example since it will always fail in submission. Bridge hub can't receive any token beyond DOT.

IkerAlus replied:
In general, system parachains will not be able to receive foreign assets.

IkerAlus reviewed (approved):
Superb work, I only have a comment about an example. The rest looks good to me. 

PS: I found some not accurate error messages while testing different inputs, but I will open a different issue for them since they are not directly related to this PR.

acatangiu commented on examples/kusama/assetHub/bridgeTransfers/assetHubDOTToPolkadot.ts:
```diff
@@ -0,0 +1,47 @@
+	let callInfo: TxResult<'call'>;
+	try {
+		callInfo = await assetApi.createTransferTransaction(
+			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Polkadot"}}}`,
```
The API is weird to assume Polkadot **Asset Hub** when the `destChainId: string` provided here is an XCM Location defining Polkadot **Relay Chain**.

PAH should be correctly expressed as:
```diff
-			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Polkadot"}}}`,
+ 			`{"parents":"2","interior":{"X2":{"GlobalConsensus":"Polkadot", "Parachain": 1000}}}`,
```

IkerAlus replied:
good catch, the API shouldn't make this assumption.

marshacb replied:
Great point. I'll update the API to be based on the location specifically passed and remove any assumptions around AssetHub.

acatangiu commented on examples/kusama/assetHub/bridgeTransfers/assetHubKSMToPolkadot.ts:
```diff
@@ -0,0 +1,47 @@
+	let callInfo: TxResult<'call'>;
+	try {
+		callInfo = await assetApi.createTransferTransaction(
+			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Polkadot"}}}`,
```
```diff
-			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Polkadot"}}}`,
+ 			`{"parents":"2","interior":{"X2":{"GlobalConsensus":"Polkadot", "Parachain": 1000}}}`,
```

acatangiu commented on examples/polkadot/assetHub/bridgeTransfers/assetHubDOTToKusama.ts:
```diff
@@ -0,0 +1,47 @@
+	let callInfo: TxResult<'call'>;
+	try {
+		callInfo = await assetApi.createTransferTransaction(
+			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Kusama"}}}`,
```
```diff
-			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Kusama"}}}`,
+ 			`{"parents":"2","interior":{"X2":{"GlobalConsensus":"Kusama", "Parachain": 1000}}}`,
```

acatangiu commented on examples/polkadot/assetHub/bridgeTransfers/assetHubKSMToKusama.ts:
```diff
@@ -0,0 +1,47 @@
+	let callInfo: TxResult<'call'>;
+	try {
+		callInfo = await assetApi.createTransferTransaction(
+			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Kusama"}}}`,
```
```diff
-			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Kusama"}}}`,
+ 			`{"parents":"2","interior":{"X2":{"GlobalConsensus":"Kusama", "Parachain": 1000}}}`,
```

acatangiu commented on examples/rococo/assetHub/bridgeTransfers/assetHubROCToWestend.ts:
```diff
@@ -0,0 +1,47 @@
+	let callInfo: TxResult<'call'>;
+	try {
+		callInfo = await assetApi.createTransferTransaction(
+			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Westend"}}}`,
```
```diff
-			`{"parents":"2","interior":{"X1":{"GlobalConsensus":"Westend"}}}`,
+ 			`{"parents":"2","interior":{"X2":{"GlobalConsensus":"Westend", "Parachain": 1000}}}`,
```

acatangiu commented on examples/rococo/assetHub/foreignAssetTransfers/reserve/assetHubMUSEToBifrostRemoteReserve.ts:
```diff
@@ -0,0 +1,59 @@
+				format: 'call',
+				xcmVersion: 4,
+				paysWithFeeDest: '{"parents":"1","interior":{"X1":{"Parachain":"3369"}}}',
+				assetTransferType: 'RemoteReserve',
+				remoteReserveAssetTransferTypeLocation: '{"parents":"1","interior":{"X1":{"Parachain":"3369"}}}',
```
When the remote reserve used is the same as asset "origin/id", then `pallet_xcm::transferAssets` also works, no need to use the more complicated `transferAssetsUsingTypeAndThen`.

`transferAssetsUsingTypeAndThen` is required when using "other" reserves (such as AssetHub) or when needing to do something else with the asset once it arrives (specify xcm to run on final dest).

IkerAlus replied:
Probably this example was meant to show how to do a remote reserve transfer using the new `transferAssetsUsingTypeAndThen`, but I am ok with simplifying it with the "default case". 

Also, as commented [here](https://github.com/paritytech/asset-transfer-api/issues/400#issuecomment-2102235785), this PR gives initial support to the new call. We should add meaningful examples for the complex use cases once they are nicely supported by the API.

acatangiu replied:
yes, I don't mean you should change it, just wanted to share the information :)

acatangiu commented on examples/rococo/assetHub/foreignAssetTransfers/reserve/assetHubMUSEToRhala.ts:
```diff
@@ -0,0 +1,52 @@
+import { GREEN, PURPLE, RESET } from '../../../../colors';
+
+/**
+ * In this example we are creating a `polkadotXcm` pallet `transferAssetsUsingTypeAndThen` call to send MUSE (foreign asset with location `{"parents":"1","interior":{"X1":{"Parachain":"3369"}}}`)
+ * from a Rococo Asset Hub (System Parachain) account
+ * to a Rhala Testnet (ParaChain) account, where the `xcmVersion` is set to 4, and there is no
+ * `weightLimit` option provided which declares that the tx will allow unlimited weight to be used for fees.
```
yes, here we need to use the `transferAssetsUsingTypeAndThen` in order to "force" using Asset Hub as `LocalReserve` 👍

acatangiu commented on examples/rococo/assetHub/foreignAssetTransfers/reserve/assetHubWETHToRhala.ts:
```diff
@@ -0,0 +1,51 @@
+import { GREEN, PURPLE, RESET } from '../../../../colors';
+
+/**
+ * In this example we are creating a `polkadotXcm` pallet `transferAssets` call to send WETH (foreign asset with location `{"parents":"2","interior":{"X2":[{"GlobalConsensus":{"Ethereum":{"chainId":"11155111"}}},{"AccountKey20":{"network":null,"key":"0xfff9976782d46cc05630d1f6ebab18b2324d6b14"}}]}}`)
+ * from a Rococo Asset Hub (System Parachain) account
+ * to a Rhala Testnet (ParaChain) account, where the `xcmVersion` is set to 4, and there is no
```
This will not work (make sure to actually validate these on Rococo).

`transferAssets` uses the asset ID to determine its reserve => in this case it will try to use the Ethereum network as the reserve and try to route the transfer through there and will fail because Ethereum is unreachable/across the bridge.

In this case you want AssetHub to act as a LocalReserve -> use `transferAssetsUsingTypeAndThen` just like you did in `assetHubMUSEToRhala.ts`.

Make sure to actually validate on Rococo all of the calls generated by these APIs!

marshacb replied:
Hey Adrian, thanks for adding this additional context! When testing submitting this tx in particular, it appeared to succeed from the perspective of AH so the assumption I made was based on that. I'll be sure to update this example to use `LocalReserve` as you mentioned.

acatangiu replied:
Make sure to check for the intended side-effects as well (transfer reaches and correctly handled on receiving side). It can happen that something works on the sending side but will not work on the recipient side.

acatangiu reviewed (commented):
Very nice!

Left some comments for some corrections.
